internal/controller/user_controller.go

```diff
@@ -0,0 +1,769 @@
+			log.Error(err, "Failed to update status regarding reconciliation status.")
+			return ctrl.Result{Requeue: true}, fmt.Errorf("failed to update status regarding reconciliation status: %w", err)
+		}
+		return ctrl.Result{}, nil
+	}
+}
+
+func getNamespace(u v1.ObjectMeta, ref solrv1alpha1.ObjectRef) string {
+	if ref.Namespace != "" {
+		return ref.Namespace
+	} else {
+		return u.Namespace
+	}
+}
+
+func (r *UserReconciler) getSecret(ctx context.Context, user *solrv1alpha1.User, try_create bool) (*corev1.Secret, error) {
+	log := logf.FromContext(ctx)
+	ref := user.Spec.Secret.ToObjectKey()
+	ref.Namespace = getNamespace(user.ObjectMeta, user.Spec.Secret.ObjectRef)
+
+	var secret corev1.Secret
+	if err := r.Get(ctx, ref, &secret); err != nil {
+		meta.SetStatusCondition(&user.Status.Conditions, v1.Condition{
+			Type:    conditionSecretExists,
+			Status:  v1.ConditionUnknown,
+			Reason:  "FailedLoad",
+			Message: fmt.Sprintf("Failed to load secret: %v", err),
+		})
+
+		if try_create {
+			secret = corev1.Secret{
+				ObjectMeta: v1.ObjectMeta{
+					Name:      ref.Name,
+					Namespace: ref.Namespace,
+				},
+			}
+			if err := controllerutil.SetOwnerReference(user, &secret, r.Scheme); err != nil {
+				log.Error(err, "Failed to set owner of created secret.")
+			}
+
+			if err := r.Create(ctx, &secret); err != nil {
+				meta.SetStatusCondition(&user.Status.Conditions, v1.Condition{
+					Type:    conditionSecretExists,
+					Status:  v1.ConditionUnknown,
+					Reason:  "FailedCreation",
+					Message: fmt.Sprintf("Failed to create secret: %v", err),
+				})
+				return nil, fmt.Errorf("failed to create secret: %w", err)
+			}
+			meta.SetStatusCondition(&user.Status.Conditions, v1.Condition{
+				Type:    conditionSecretExists,
+				Status:  v1.ConditionTrue,
+				Reason:  "Created",
+				Message: "Created secret.",
+			})
+			for _, value := range []string{conditionSecretUsername, conditionSecretPassword, conditionSecretEndpoint, conditionSecretEndpointHostname} {
+				meta.SetStatusCondition(&user.Status.Conditions, v1.Condition{
+					Type:    value,
+					Status:  v1.ConditionFalse,
+					Reason:  "Created",
+					Message: "Empty secret created.",
+				})
+			}
+
+			log.Info("Created secret",
+				"namespace", ref.Namespace,
+				"name", ref.Name,
+			)
+			return &secret, nil
+		}
+		return nil, fmt.Errorf("failed to load secret for %v, %s: %w", user, ref, err)
+	}
+	meta.SetStatusCondition(&user.Status.Conditions, v1.Condition{
+		Type:    conditionSecretExists,
+		Status:  v1.ConditionTrue,
+		Reason:  "Exists",
+		Message: "Secret exists.",
+	})
+	return &secret, nil
+}
+
+type Credentials struct {
+	Username string `spec_field_name:"UsernameKey" condition:"SecretUsername"`
+	Password string `spec_field_name:"PasswordKey" condition:"SecretPassword"`
+	Endpoint string `spec_field_name:"EndpointKey" condition:"SecretEndpoint"`
+	Hostname string `spec_field_name:"EndpointHostnameKey" condition:"SecretEndpointHostname"`
```

Comment:
Unsure where best to comment on this, but: Moved away from using this, in the end... could cut it out, but given time constraints, wanting to say we can live with it existing for now.